### ISSUE TITLE
Relation between Music view and Manuscript viewer

### ISSUE THREAD

**Issue by uliska** [wbsoft/frescobaldi#746]
This is related to #737 but also something different.
I add this to the "initial manuscript viewer" milestone but intend to move it along to the next one.

I realized that the manuscript viewer in the form of our intended initial release is about 90% the same as the music viewer, so it will be perfectly possible to rebuild the music viewer as a copy of the manuscript viewer. (And I'm tempted to say we should try to do it for the initial release, but I know there will be much more details than I expect now, so I know it's not a good idea ...). The main difference is the handling of documents.

I'll go through the different aspects to elaborate on the identities and differences. This is not *completely* exhaustive, there are details that can be discussed when time comes:

### Display

Both viewers display a "document" built from individual pages that are currently taken from a PDF document. @wbsoft promises to extend the poppler view infrastructure so these pages can also use other image formats. This will support using manuscripts that are actually series of image files (jpg, tif, png?), which will then behave identically.
Maybe this will then make it possible to also display the compiled results when compiling to PNG.

An open question is if it will be possible to also integrate SVG images in that structure. It would be *very* desirable to have the SVG viewer be part of the inheritance hierarchy to provide a consistent interface to it. If the poppler view can't support that we should look into replacing the `View` part in the current set-up with a custom view based on `QWebView` but still have the panel in the unified infrastructure.

### Toolbar

The musicview should get the same toolbar as the new manuscript viewer's. Having a hidable toolbar withing the panel is consistent and probably even better than the current implementation where the (visual/psychological) link between the panel and the toolbar is quite weak.
The new implementation makes it very straightforward to customize the different viewers' toolbars.

### Context menu

The musicview should also get the new context menu. It is more powerful and extensible and - as with the toolbar - it is straightforward to customize for a concrete viewer's needs.

##### Open/Close actions

It has to be discussed if these actions should be suppressed from the music view. Presumably it'll be best to keep the behaviour as it is now, making the loading of documents completely depending on the actions in the input editor. However, they are there ...

##### Print, Zoom and Pager actions

They are identical, so it is good to have them all implemented in the common base class. The same is true with the magnifying glass
.
##### Synchronize view, point-and-click

Should be the same. So far they only work in the manuscript viewer when a LilyPond generated PDF is displayed, but we will later add support for another linking system for arbitrary manuscripts.
So the actions should be defined in the base class, and later we'll be able to override the necessary behaviour.

### Document Chooser

This is the item with the most notable differences. Actually the handling of documents is the most or even single fundamental difference between music view and manuscript view.

The music view automatically opens a (list of) PDF document(s) depending on the current *input document* while the manuscript viewer maintains an arbitrary list of manually opened files. However, this will change when we add support for manuscripts defined in a helper file so they are also partially dependent on the input document.

I think (and this relates to #737) the responsibility to maintain (and perhaps cache) the opened documents should be moved to the viewer's base class (not sure whether this should be the panel, the widget or the view, though). Then, what now is the document chooser will concentrate on the differences, that is in populating the list of files that is managed ii the viewer.
This should make a cleaner design IISC.

### Session handling

This is related to the document chooser.
I think this should also be made more consistent in so far as the music view should also save its opened files in a session. So far this is not the case, and the files are implicitly handled upon the changes in the input documents. Sharing the session handling code (which it would do automatically BTW) will add the feature of restoring positions to the music view.

One thing we will have to take care of (which is currently done somewhere else) is checking if a generated PDF is up-to-date with the input document.
OTOH the music view can simply ignore missing PDFs as they will be generated by LilyPond. Fortunately this is as easy as overriding the `reportMissingDocuments` method with a `pass` method.

**Comment by PeterBjuhr** [comment]
>I think this should also be made more consistent in so far as the music view should also save its opened files in a session. 

Is there any need for that? The pdf:s are semi-restored from the editor documents.

>Sharing the session handling code will add the feature of restoring positions to the music view.

Yes, that is nice of course. But what is the use case, proof-reading?

**Comment by uliska** [comment]
>> I think this should also be made more consistent in so far as the music view should also save its opened files in a session.

> Is there any need for that? The pdf:s are semi-restored from the editor documents.

No real need, just in the context of the other aspect below.

>>  Sharing the session handling code will add the feature of restoring positions to the music view.

> Yes, that is nice of course. But what is the use case, proof-reading?

It's more about consistency, to come back to the position where you left.

**Comment by uliska** [comment]
The cursor position in the input document is restored too, so it makes sense if the displayed score is also restored at the previous position.

**Comment by uliska** [comment]
I think my latest commits show that the approach is quite promising. If you look at the definitions of the `viewers.manuscript.widget.ManuscriptViewWidget` and `viewers.manuscript.ManuscriptViewPanel` classes I think it becomes apparent that the child classes only need to contain the little bit of individualization.
Once the manuscript viewer will be developed further the differences will grow, but we'll know what code has to go into the concrete manuscript viewer's classes.

What I haven't touched so far is the whole DocumentChooser part. But maybe we'll leave this out for later. But eventually I think this will lead to a similar clean-up.

**Comment by PeterBjuhr** [comment]
>It's more about consistency, to come back to the position where you left.

I have nothing against the idea of saving the positions (and it is very useful for the manuscript viewer). I just want to make sure that we are not storing unnecessary data with the session. What I can recall none has expressed a wish for such a feature. That doesn't mean it can't be useful of course...

**Comment by uliska** [comment]
No, there's no explicit need for that function. But IISC we'll get it automatically and would otherwise have to explicitly suppress it.

**Comment by uliska** [comment]
By now (c474727f) I have cleaned up all of the files inside `viewers.manuscript`. They are extremely concise now, showing that my above assumptions are correct. The music view and the manuscript view share extremely much, and I have moved nearly everything to the common base class. It will be interesting to redo the music view from scratch as a new child to our abstract viewer.
If I'm not mistaken the only differences will be in the handling of changes to the editor documents, i.e. the automatic retrieval of viewer documents.

**Comment by uliska** [comment]
I think this is ready by now, but I'll move it to the next milestone.
